File: packages/vahtian-py/tests/test_core.py
import vahtian

# The cross-language parity gate: the R package `vahtian` asserts this SAME literal.
# If either canonical serialiser drifts, one of the two CIs goes red.
GOLDEN = "sha256:50ca741a72e7058870d0ca7594b0c37faa7183472fcb1752b7a6c5abe23cafd1"

def test_golden_cross_language_hash():
    recs = [
        {"pmid": "12345", "title": "PD-L1 AI scoring agrees with pathologists",
         "provenance": [{"source": "pubmed", "retrieved": "2026-06-23"}]},
        {"doi": "10.1/x", "title": "A second study",
         "provenance": [{"source": "openalex", "retrieved": "2026-06-23"}]},
        {"pmid": "12345", "title": "PD-L1 AI scoring agrees with pathologists",
         "provenance": [{"source": "europepmc", "retrieved": "2026-06-23"}]},
    ]
    c = vahtian.freeze(recs, search_date="2026-06-23")
    assert len(c.records) == 2
    assert c.content_hash == GOLDEN

def test_dedupe_and_reproducible():
    recs = [
        {"pmid": "1", "title": "A", "provenance": [{"source": "pubmed"}]},
        {"pmid": "1", "title": "A", "provenance": [{"source": "europepmc"}]},
        {"doi": "10.1/x", "title": "B", "provenance": [{"source": "openalex"}]},
    ]
    c1 = vahtian.freeze(recs, search_date="2026-06-23", now="t")
    c2 = vahtian.freeze(list(reversed(recs)), search_date="2026-06-23", now="t2")
    assert len(c1.records) == 2
    assert c1.content_hash == c2.content_hash   # order-independent
    assert c1.verify()

def test_tamper_detection():
    c = vahtian.freeze([{"pmid": "1", "title": "A"}], search_date="2026-06-23")
    assert c.verify()
    c.records[0]["title"] = "tampered"
    assert not c.verify()

def test_audit_chain():
    L = vahtian.Ledger()
    L.append("human", "rate", {"v": "supported"}, ts="t1")
    L.append("ai", "advise", {"v": "supported"}, ts="t2")
    assert L.verify()
    L.entries[0]["payload"]["v"] = "x"
    assert not L.verify()

File: packages/vahtian-r/tests/testthat/test-core.R
test_that("golden cross-language hash matches the Python package byte-for-byte", {
  # Same fixture as the Python test; the GOLDEN hash was computed by `vahtian` (Python).
  recs <- list(
    list(pmid = "12345", title = "PD-L1 AI scoring agrees with pathologists",
         provenance = list(list(source = "pubmed", retrieved = "2026-06-23"))),
    list(doi = "10.1/x", title = "A second study",
         provenance = list(list(source = "openalex", retrieved = "2026-06-23"))),
    list(pmid = "12345", title = "PD-L1 AI scoring agrees with pathologists",
         provenance = list(list(source = "europepmc", retrieved = "2026-06-23")))
  )
  corpus <- vahtian_freeze(recs, search_date = "2026-06-23", now = "fixed")
  expect_equal(length(corpus$records), 2L)              # deduped by record id
  expect_identical(
    corpus$content_hash,
    "sha256:50ca741a72e7058870d0ca7594b0c37faa7183472fcb1752b7a6c5abe23cafd1"
  )
  expect_true(vahtian_verify(corpus))
})

test_that("reproducible independent of source order", {
  recs <- list(
    list(pmid = "1", title = "A", provenance = list(list(source = "pubmed"))),
    list(pmid = "1", title = "A", provenance = list(list(source = "europepmc")))
  )
  a <- vahtian_freeze(recs, search_date = "2026-06-23", now = "t")
  b <- vahtian_freeze(rev(recs), search_date = "2026-06-23", now = "t2")
  expect_identical(a$content_hash, b$content_hash)
})

test_that("tamper detection", {
  corpus <- vahtian_freeze(list(list(pmid = "1", title = "A")), search_date = "2026-06-23")
  expect_true(vahtian_verify(corpus))
  corpus$records[[1]]$title <- "tampered"
  expect_false(vahtian_verify(corpus))
})

test_that("audit chain detects retro-edits", {
  L <- ledger()
  L <- ledger_append(L, "human", "rate", list(v = "supported"), ts = "t1")
  L <- ledger_append(L, "ai", "advise", list(v = "supported"), ts = "t2")
  expect_true(ledger_verify(L))
  L$entries[[1]]$payload$v <- "x"
  expect_false(ledger_verify(L))
})
